src/ray/gcs/gcs_server/gcs_placement_group_scheduler.cc

```diff
@@ -0,0 +1,244 @@
+                 << bundle.BundleId();
+  auto status = lease_client->RequestResourceReserve(
+      bundle, [this, node_id, bundle, node, callback](
+                  const Status &status, const rpc::RequestResourceReserveReply &reply) {
+        // If the bundle is still in the leasing map and the status is ok, remove the
+        // bundle from the leasing map and handle the reply. Otherwise, lease again,
+        // because it may be a network exception. If the bundle is not in the leasing
+        // map, it means that the placement_group has been cancelled as the node is
+        // dead, just do nothing in this case because the gcs_placement_group_manager
+        // will reconstruct it again.
+        // TODO(AlisaWu): Add placement group cancel.
+        auto iter = node_to_bundles_when_leasing_.find(node_id);
+        if (iter != node_to_bundles_when_leasing_.end()) {
+          auto bundle_iter = iter->second.find(bundle.BundleId());
+          RAY_CHECK(bundle_iter != iter->second.end());
+          if (status.ok()) {
+            if (reply.success() == true) {
+              // Remove the actor from the leasing map as the reply is returned from the
+              // remote node.
+              iter->second.erase(bundle_iter);
+              if (iter->second.empty()) {
+                node_to_bundles_when_leasing_.erase(iter);
+              }
+              RAY_LOG(DEBUG) << "Finished leasing resource from " << node_id
+                             << " for bundle " << bundle.BundleId();
+            } else {
+              RAY_LOG(DEBUG) << "Failed leasing resource from " << node_id
+                             << " for bundle " << bundle.BundleId();
+            }
+          }
+          callback(status, reply);
+        }
+      });
+  if (!status.ok()) {
+    rpc::RequestResourceReserveReply reply;
+    reply.set_success(false);
+    callback(Status::OK(), reply);
+  }
+}
+
+void GcsPlacementGroupScheduler::CancelResourceReserve(
+    BundleSpecification bundle_spec, std::shared_ptr<ray::rpc::GcsNodeInfo> node) {
+  RAY_CHECK(node);
+
+  auto node_id = ClientID::FromBinary(node->node_id());
+  RAY_LOG(DEBUG) << "Start returning resource for node " << node_id << " for bundle "
+                 << bundle_spec.BundleId();
+
+  rpc::Address remote_address;
+  remote_address.set_raylet_id(node->node_id());
+  remote_address.set_ip_address(node->node_manager_address());
+  remote_address.set_port(node->node_manager_port());
+  auto return_client = GetOrConnectLeaseClient(remote_address);
+  auto status = return_client->CancelResourceReserve(
+      bundle_spec,
+      [this, bundle_spec, node](const Status &status,
+                                const rpc::CancelResourceReserveReply &reply) {
+        if (!status.ok()) {
+          // TODO(AlisaWu): add a wait time.
+          CancelResourceReserve(bundle_spec, node);
+        }
+      }
+
+  );
+  if (!status.ok()) {
```

Comment:
> Is this necessary? When is the case `status` is not `ok`?
maybe timeout or rpc error?

Reply:
Ah, it makes sense. So you are retrying it whenever there's any type of failure.

Reply:
maybe timeout or rpc error? => Doesn't this mean it is highly likely that the node is dead? Isn't there any possibility this will cause infinite recursion?

Reply:
Seems like the same mechanism is used for leasing workers, so I guess there's some mechanism in there?